Feature/pyro visualization

matt-edminster opened the pull request:
- Added a List of Pyro Cues to the metadata screen,

- Added visualization around the UAV in the viewer when pyro is triggered, stay on for the duration of firing

- Added pyro firing events to the timeline slider with labels

matt-edminster committed:
feat: add pyro event inspection and visualization

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/skybrush-io/viewer?pullRequest=17) <br/>All committers have signed the CLA.

isti115 committed:
fix: restore vertical alignment of playback slider

isti115 committed:
chore: apply automatic formatting using prettier

isti115 committed:
chore: clean up unused imports and parameters

isti115 committed:
refactor: extract pyro related constants

isti115 committed:
refactor: improve code style using modern features

Prefer `find`, `flatMap`, `some`, `toSorted` and such over nested loops
with mutable variables combined with `break` and `continue` statements.

isti115 committed:
refactor: remove reduntant pyro program validation

isti115 committed:
chore: fix types in `drone-flock.ts` after rebase

isti115 committed:
chore: update `@skybrush/show-format`

Version 5.1.0 exports `PyroProgram` and adds a validator function.

isti115 committed:
feat: remove broken code for always visible labels

isti115 committed:
feat: enable click to jump on the pyro cue sheet

Allow navigation from the items to their respective timestamps.

isti115 committed:
feat: add dedicated shader for pyro visualization

isti115 commented:
Thanks for the contribution and sorry for the delay, some major repository maintenance works have begun just as I was picking this up again, which I didn't want to merge into the middle of, but now that is over, so I have rebased and cleaned up the suggested changes.

I have done quite extensive refactoring as well as making some changes in terms of functionality, I hope that you'll like the results! 🙂

You can find the latest revision pushed back to the original `feature/pyro-visualization` branch if you would like to test it. There still are some TODOs here and there (e.g. missing switch for toggling pyro visualization on and off), but I think it's valuable to get feedback from @ntamas and @vasarhelyi at this stage before moving forward.

I might have gone a bit overboard regarding the actual appearance of the "indicators", so it's possible that we will have to dial it back a bit. 😀 I took the artistic liberty of creating a custom shader for the pyro entity, but I have also added an indicator octahedron that's protruding out of the spheres along all axes in case it becomes difficult to tell which object the sparkles are falling out of. That can be seen from all angles and has an animated material in order to stay visible no matter what color the sphere is.

All in all, I consider this ready to merge in its current state, what's missing is for you to sign the CLA and then we can keep iterating on the main branch or in further pull requests.

matt-edminster commented:
Sounds good will test out more soon but just signed the CLA

ntamas commented on src/features/show/selectors.ts:
```diff
@@ -343,14 +349,142 @@ export const getLightProgramPlayers = createSelector(
+      // Truncate long payload names for timeline labels
+      // TODO: Use the proper ellipsis character: "…"
+      if (payloadText.length > 25) {
+        payloadText = payloadText.substring(0, 22) + '...';
```
This would have probably been easier to fix directly than to add the proper ellipsis character in the _comment_ :-D

ntamas commented on src/features/show/selectors.ts:
```diff
@@ -343,14 +349,142 @@ export const getLightProgramPlayers = createSelector(
+
+      // Truncate long payload names for timeline labels
+      // TODO: Use the proper ellipsis character: "…"
+      if (payloadText.length > 25) {
```
Probably make the truncation length an external constant that is referenced here.

ntamas commented on src/views/player/sidebar/MetadataSection.tsx:
```diff
@@ -80,10 +87,55 @@ export default function MetadataSection() {
+            const payloadText = payloadNames.join(', ');
+            // Truncate long payload names to fit better
+            const truncatedPayload =
+              payloadText.length > 25
```
Anotehr appearance of the magic constant for the truncation length.

ntamas commented on src/views/player/sidebar/MetadataSection.tsx:
```diff
@@ -80,10 +87,55 @@ export default function MetadataSection() {
+            // Truncate long payload names to fit better
+            const truncatedPayload =
+              payloadText.length > 25
+                ? payloadText.substring(0, 24) + '…'
```
Now we have the proper ellipsis character here! :)

ntamas reviewed (approved):
LGTM; see a few comments for suggestions.

ntamas commented:
I'll keep on iterating over this a bit but it looks great in general!
